backtracking/sudoku.py

```diff
@@ -0,0 +1,127 @@
+
+'''
+
+	Given a partially filled 9×9 2D array, the objective is to fill a 9×9 
+	square grid with digits numbered 1 to 9, so that every row, column, and 
+	and each of the nine 3×3 sub-grids contains all of the digits.
+
+	This can be solved using Backtracking and is similar to n-queens.
+	We check to see if a cell is safe or not and recursively call the 
+	function on the next column to see if it returns True. if yes, we
+	have solved the puzzle. else, we backtrack and place another number
+	in that cell and repeat this process. 
+
+'''
+
+def is_safe(grid, row, column, n):
+	'''
+	This function checks the grid to see if each row,
+	column, and the 3x3 subgrids contain the digit 'n'.
+	It returns False if it is not 'safe' (a duplicate digit
+	is found) else returns True if it is 'safe'
+	
+	'''
+
+	for i in range(9):
+		if(grid[row][i] == n):
+			return False
+		
+	for i in range(9):
+		if(grid[i][column] == n):
+			return False		
+	
+	for i in range(3):
+		for j in range(3):
+			if(grid[(row - row%3) + i][(column - column%3) + j] == n):
+				return False
+
+	return True			
+
+
+def is_completed(grid):
+	'''
+	This function checks if the puzzle is completed or not. 
+	it is completed when all the cells are assigned with a number(not zero)
+	and There is no repeating number in any column, row or 3x3 subgrid.
+	
+	'''
+	
+	for i in range(9):
+		for j in range(9):
+			if(grid[i][j] == 0):
+				return False
+
+	return True
```

Comment:
OPTIONAL: Challenges:
1. Try to do this without hardcoding 9 and without using __range()__ like: __for row in grid: for cell in row:__
2.Try making the body of this function a one-liner using nested list comprehensions.